src/http/h1/decode.rs

```diff
@@ -136,113 +126,177 @@ impl Decoder {
-            b';' if !in_ext => {
-                in_ext = true;
-                in_chunk_size = false;
-            },
-            // "Linear white space" is ignored between the chunk size and the
-            // extension separator token (";") due to the "implied *LWS rule".
-            b'\t' | b' ' if !in_ext & !in_chunk_size => {},
-            // LWS can follow the chunk size, but no more digits can come
-            b'\t' | b' ' if in_chunk_size => in_chunk_size = false,
-            // We allow any arbitrary octet once we are in the extension, since
-            // they all get ignored anyway. According to the HTTP spec, valid
-            // extensions would have a more strict syntax:
-            //     (token ["=" (token | quoted-string)])
-            // but we gain nothing by rejecting an otherwise valid chunk size.
-            _ext if in_ext => {
-                //TODO: chunk extension byte;
-            },
-            // Finally, if we aren't in the extension and we're reading any
-            // other octet, the chunk size line is invalid!
+            b'\t' | b' ' => return Ok(ChunkedState::SizeLws),
+            b'\r' => return Ok(ChunkedState::SizeLf),
             _ => {
                 return Err(io::Error::new(io::ErrorKind::InvalidInput,
-                                         "Invalid chunk size line"));
+                                          "Invalid chunk size line: Invalid Size"));
             }
         }
+        Ok(ChunkedState::Size)
     }
-    trace!("chunk size={:?}", size);
-    Ok(size)
-}
+    fn read_size_lws<R: Read>(rdr: &mut R) -> io::Result<ChunkedState> {
+        trace!("read_size_lws");
+        match byte!(rdr) {
+            // LWS can follow the chunk size, but no more digits can come
+            b'\t' | b' ' => Ok(ChunkedState::SizeLws),
+            b'\r' => return Ok(ChunkedState::SizeLf),
+            _ => Err(io::Error::new(io::ErrorKind::InvalidInput,
+                                           "Invalid chunk size linear white space")),
+        }
+    }
+    fn read_size_lf<R: Read>(rdr: &mut R, size: &mut usize) -> io::Result<ChunkedState> {
+        trace!("Chunk size is {:?}", size);
+        match byte!(rdr) {
+            b'\n' if *size > 0 => Ok(ChunkedState::Body),
+            b'\n' if *size == 0 => Ok(ChunkedState::End),
+            _ => Err(io::Error::new(io::ErrorKind::InvalidInput,
+                                           "Invalid chunk size LF")),
+        }
+    }
+    fn read_body<R: Read>(rdr: &mut R, rem: &mut usize, buf: &mut [u8], read: &mut usize)
+            -> io::Result<ChunkedState> {
+        trace!("Chunked read, remaining={:?}", rem);
+
+        let to_read = cmp::min(*rem, buf.len());
+        let count = try!(rdr.read(&mut buf[..to_read]));
 
+        trace!("to_read = {}", to_read);
+        trace!("count = {}", count);
+
+        if count == 0 {
+            *rem = 0;
+            return Err(io::Error::new(io::ErrorKind::Other, "early eof"));
```

Comment:
It didn't exist when this code was originally written (pre Rust 1.0), but there is now `io::ErrorKind::UnexpectedEof`, which seems more appropriate here.